benches/latency.rs

```diff
@@ -0,0 +1,218 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::net::SocketAddr;
+
+use futures::{Future, Poll};
+use tokio_core::net::UdpSocket;
+use tokio_core::reactor::Core;
+
+use test::Bencher;
+use std::thread;
+use std::time::Duration;
+
+use futures::stream::Stream;
+
+/// UDP echo server
+struct EchoServer {
+    socket : UdpSocket,
+    buf : Vec<u8>,
+    to_send : Option<(usize, SocketAddr)>,
+    stop : Arc<AtomicBool>,
+}
+
+impl EchoServer {
+    fn new(s : UdpSocket, stop : Arc<AtomicBool>) -> Self {
+        EchoServer {
+            socket: s,
+            to_send: None,
+            buf: vec![0u8; 1600],
+            stop: stop,
+        }
+    }
+}
+
+impl Future for EchoServer {
+    type Item = ();
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<(), io::Error> {
+        loop {
+            if self.stop.load(Ordering::SeqCst) {
+                return Ok(futures::Async::Ready(()))
+            }
+
+            if let Some((size, peer)) = self.to_send.take() {
+                match self.socket.send_to(&self.buf[..size], &peer) {
+                    Err(e) => {
+                        self.to_send = Some((size, peer));
+                        return try_nb!(Err(e));
+                    },
+                    Ok(_) => { }
+                }
+            }
+
+            self.to_send = Some(
+                try_nb!(self.socket.recv_from(&mut self.buf))
+                );
+        }
+    }
+}
+
+/// UDP echo server
+struct ChanServer {
+    rx : tokio_core::channel::Receiver<usize>,
+    tx : tokio_core::channel::Sender<usize>,
+    buf : Option<usize>,
+    stop : Arc<AtomicBool>,
+}
+
+impl ChanServer {
+    fn new(tx : tokio_core::channel::Sender<usize>, rx : tokio_core::channel::Receiver<usize>, stop : Arc<AtomicBool>) -> Self {
+        ChanServer {
+            rx: rx,
+            tx: tx,
+            buf: None,
+            stop: stop,
+        }
+    }
+}
+
+impl Future for ChanServer {
+    type Item = ();
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<(), io::Error> {
+        loop {
+            if self.stop.load(Ordering::SeqCst) {
+                return Ok(futures::Async::Ready(()))
+            }
+
+            if let Some(u) = self.buf.take() {
+                match self.tx.send(u) {
+                    Err(e) => {
+                        self.buf = Some(u);
+                        return try_nb!(Err(e));
+                    },
+                    Ok(_) => { }
+                }
+            }
+
+            match self.rx.poll() {
+                Ok(futures::Async::Ready(None)) => return Ok(futures::Async::Ready(())),
+                Ok(futures::Async::Ready(Some(t))) => { self.buf = Some(t) },
+                Ok(futures::Async::NotReady) => return Ok(futures::Async::NotReady),
+                Err(e) => return try_nb!(Err(e)),
+            }
+        }
+    }
+}
+#[bench]
+fn udp_echo_latency(b: &mut Bencher) {
+    let server_addr= "127.0.0.1:7398".to_string();
```

Comment:
Sure but a message needs to be sent anyway as soon as the socket is bound, so this'd just be precisely that message